backend/core/src/shared/uploads/aws/amazon-s3-file-service.ts

```diff
@@ -0,0 +1,106 @@
+import { BaseFileService } from "../base-file-service"
+import { ConfigInvalidEnumError, ConfigItemMissingError } from "../errors"
+import { FileUpload, FileServiceConfig, FileService } from "../types"
+import { S3Client, PutObjectCommand, GetObjectCommand } from "@aws-sdk/client-s3"
+import { getSignedUrl } from "@aws-sdk/s3-request-presigner"
+
+enum UrlFormats {
+  public = "public",
+  private = "private",
+}
+
+export class AmazonS3FileService extends BaseFileService implements FileService {
+  client: S3Client
+  region: string
+  bucket: string
+  urlFormat: string
+
+  validateConfig(config: FileServiceConfig) {
+    if (!config.region) {
+      throw new ConfigItemMissingError("region")
+    }
+
+    if (!config.bucket) {
+      throw new ConfigItemMissingError("bucket")
+    }
+
+    // enums enumerate values weirdly, so do it by key
+    const validUrlFormats = Object.keys(UrlFormats).map((key) => UrlFormats[key])
+
+    if (!validUrlFormats.includes(config.url_format)) {
+      throw new ConfigInvalidEnumError("url_format", validUrlFormats)
+    }
+  }
+
+  configure(config: FileServiceConfig) {
+    super.configure(config)
+
+    // set local props based on validated config items
+    this.client = new S3Client({
+      region: config.region,
+    })
+
+    this.region = config.region
+    this.bucket = config.bucket
+    this.urlFormat = config.url_format
+  }
+
+  private async _generateUrl(region: string, bucket: string, key: string) {
```

Comment:
I personally had issues when making this method async (which is needed for the signed URL part) since the partners code calling this expects a string and ran into issues with React when changing it. Just leaving this comment as a heads up, might make sense to remove the private part if we don't have that yet. Feel free to resolve this comment.

Reply:
Thanks for the heads up.  I suspect it's quite a bit easier to get it working when React isn't involved.  This method is tested and functional.  It's private because it is only supposed to get called by the public methods `putFile()` and `generateDownloadUrl()`.